src/libraries/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/CmsSigner.cs

```diff
@@ -188,85 +239,128 @@ internal void CheckCertificateValue()
+
+                signedAttrs.Add(
+                    new AttributeAsn
+                    {
+                        AttrType = Oids.ContentType,
+                        AttrValues = new[] { new ReadOnlyMemory<byte>(writer.Encode()) },
+                    });
             }
-            catch (PlatformNotSupportedException ex)
+            // else: TODO if we're in pure mode we *should* add a content type according to
+            // the SLH-DSA and ML-DSA spec. However, the only case when the content type is null
+            // is when we're countersigning, and RFC 5652 specifically states that
+            // countersignatures must not contain a content type. We'll leave it as is for now
+            // as countersignatures don't seem to be in the SLH-DSA CMS spec.
+
+            // Use the serializer/deserializer to DER-normalize the attribute order.
+            SignedAttributesSet signedAttrsSet = default;
+            byte[]? attributesToSign = null;
+            signedAttrsSet.SignedAttributes = PkcsHelpers.NormalizeAttributeSet(
+                signedAttrs.ToArray(),
+                encodedSignedAttrs => attributesToSign = encodedSignedAttrs);
+            Debug.Assert(
+                attributesToSign is not null,
+                $"{attributesToSign} should have been set by {nameof(PkcsHelpers.NormalizeAttributeSet)}");
+
+            // Since this contains user data in a context where BER is permitted, use BER.
+            // There shouldn't be any observable difference here between BER and DER, though,
+            // since the top level fields were written by NormalizeSet.
+            newSignedAttrsWriter = new AsnWriter(AsnEncodingRules.BER);
+            signedAttrsSet.Encode(newSignedAttrsWriter);
+
+            return attributesToSign;
+        }
+
+        internal ReadOnlyMemory<byte> GetMessageToSignPure(
+            HashAlgorithmName hashAlgorithmName,
+            ReadOnlyMemory<byte> data,
+            string? contentTypeOid,
+            out ReadOnlyMemory<byte>? signedAttributesAsn)
+        {
+            byte[] dataHash;
+            // In pure mode we will always sign the attributes rather than the message content even
+            // when signing the content is allowed. In general the attribute payload is smaller.
+            using (IncrementalHash hasher = Helpers.CreateIncrementalHash(hashAlgorithmName))
             {
-                throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmName), ex);
+                hasher.AppendData(data.Span);
+                dataHash = hasher.GetHashAndReset();
             }
 
-            using (hasher)
+            byte[] contentToSign = PrepareAttributesToSign(dataHash, contentTypeOid, out AsnWriter newSignedAttrsWriter);
+            signedAttributesAsn = newSignedAttrsWriter.Encode();
+            return contentToSign;
+        }
+
+        internal ReadOnlyMemory<byte> GetMessageToSignHashed(
```

Comment:
The "Hashed" suffix makes sense when looking at the Pure alternative, but it reads really oddly on its own.

Maybe it would make more sense as an infix, GetHashedMessageToSign?